### ISSUE TITLE
Benchmark whichmut dense vs sparse formats for re-implementation decision

### ISSUE THREAD

**Issue by jgallowa07** [matsengrp/netam#171]
## Questions to Answer

1. **Can we avoid pre-computing neutral rate tensors?**
   - Compute on-the-fly during training instead?
   - What's the performance penalty?

2. **Dense vs Sparse for re-implementation?**
   - Sparse uses ~10x less memory (9 entries vs 64 per codon)
   - But is the complexity worth it when batching?
   - Which should we recommend for a clean re-implementation?

### Test Setup

**Single realistic dataset:**
- N = 10,000 sequences
- L = 900 nucleotides (300 codons) - typical antibody length
- Batch size = 32 (typical training batch)

### Benchmark 1: On-the-fly vs Pre-computed

**Compare three approaches:**

A. **Current (pre-computed):** Store all neutral rates in memory
B. **On-the-fly:** Compute neutral rates during each forward pass
C. **Hybrid:** Cache per-sequence, recompute per-epoch

**Metrics:**
- Total memory usage
- Time per epoch
- Time for dataset creation

**Expected outcome:** Quantify the speed/memory tradeoff

```python
# Pseudo-code for on-the-fly approach
def compute_loss_with_live_rates(batch, neutral_model_fn):
    # Compute neutral rates for just this batch
    neutral_rates = neutral_model_fn(batch.sequences)
    # Rest of loss computation
    ...
```

### Benchmark 2: Dense vs Sparse (assuming pre-computed)

**Compare both formats on:**

1. **Memory footprint:**
   - Size of neutral_rates tensor/dict
   - Peak memory during training

2. **Speed:**
   - Time for 1 epoch (10,000 / 32 = 312 batches)
   - Time for single batch forward pass
   - Time for single batch backward pass

3. **Code complexity:**
   - Lines of code
   - Cyclomatic complexity
   - Ease of debugging

**Quick analysis:**
```python
# Dense memory
dense_size = N * L_codon * 65 * 4 bytes
# = 10,000 * 300 * 65 * 4 = 780 MB

# Sparse memory
sparse_size = N * L_codon * 9 * (8 + 8 + 4) bytes  # indices (2*long) + value (float)
# = 10,000 * 300 * 9 * 20 = 540 MB

# Ratio: ~1.4x, not 10x!
# Why? Sparse stores indices too (2 longs per entry)
```

### Benchmark 3: Batching Impact

**Key question:** Does batching on GPU make dense competitive?

Test with different batch sizes: [1, 8, 32, 128]
- Dense: Simple indexing, GPU-friendly
- Sparse: Search operation, harder to vectorize

**Hypothesis:** Dense may be faster on GPU despite larger memory if:
- Batching is effective
- GPU memory is sufficient
- Vectorized indexing >> sparse search

### Quick benchmark script:

```python
# netam/benchmarks/quick_whichmut_benchmark.py

import time
import torch
from memory_profiler import memory_usage

def benchmark_format(format_type, n_epochs=3):
    """Run and time training for a few epochs."""
    # Create dataset
    dataset = create_test_dataset(format=format_type)

    # Measure memory
    mem_before = torch.cuda.memory_allocated()

    # Time training
    start = time.time()
    for epoch in range(n_epochs):
        train_epoch(dataset)
    elapsed = time.time() - start

    mem_after = torch.cuda.memory_allocated()

    return {
        'memory_mb': (mem_after - mem_before) / 1e6,
        'time_per_epoch': elapsed / n_epochs,
        'format': format_type
    }

# Run both
dense_results = benchmark_format('dense')
sparse_results = benchmark_format('sparse')

# Print comparison
print(f"Dense:  {dense_results['memory_mb']:.1f} MB, {dense_results['time_per_epoch']:.2f} s/epoch")
print(f"Sparse: {sparse_results['memory_mb']:.1f} MB, {sparse_results['time_per_epoch']:.2f} s/epoch")
```

## Decision Framework

After benchmarks, decide based on:

### Choose **DENSE** if:
- Memory difference < 2x in practice
- Speed is significantly faster (>20%)
- Simpler code matters for maintainability
- Target platform has sufficient GPU memory

### Choose **SPARSE** if:
- Memory savings are substantial (>3x)
- Speed difference is negligible (<10%)
- Need to scale to larger datasets
- Running on memory-constrained hardware

### Choose **ON-THE-FLY** if:
- Memory is extremely tight
- Neutral model computation is fast
- Training is I/O bound anyway
- Dataset doesn't fit in memory

## Expected Outcomes (to validate or refute)

**Predictions:**

1. **On-the-fly will be too slow** - neutral rate computation per batch will dominate
   - Pre-computing is likely necessary

2. **Dense will be faster on GPU** - despite using more memory
   - Simple indexing vs search operation
   - Better GPU utilization with batching

3. **Memory difference will be ~2x, not 10x** - sparse has overhead
   - Indices storage is significant
   - Actual savings less than naive calculation

4. **Recommendation: DENSE for re-implementation**
   - Simpler code (easier to port/maintain)
   - Likely faster with batching
   - Memory difference not prohibitive for typical datasets
   - Only use sparse if memory is actually a bottleneck

## Timeline

- **Day 1 Morning:** Write benchmark script, run memory tests
- **Day 1 Afternoon:** Run speed tests, collect data
- **Day 2 Morning:** Analyze results, create comparison plots
- **Day 2 Afternoon:** Write up recommendation with rationale

## Deliverables

1. `netam/benchmarks/quick_whichmut_benchmark.py` - runnable script
2. `BENCHMARK_RESULTS.md` - findings summary with 2-3 plots
3. **Clear recommendation:** Dense or Sparse for re-implementation

## Related Files

- `netam/whichmut_dataset.py` - Dataset implementations
- `netam/whichmut_trainer.py` - Training and loss computation
- `netam/codon_table.py` - Codon mutation lookup tables
